homeassistant/components/opencv.py

```diff
@@ -0,0 +1,182 @@
+        vol.Required(CONF_GROUPS): vol.All(
+            cv.ensure_list,
+            [CLASSIFIER_GROUP_SCHEMA]
+        ),
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+# NOTE:
+# pylint cannot find any of the members of cv2, using disable=no-member
+# to pass linting
+
+
+def cv_image_to_bytes(cv_image):
+    """Convert OpenCV image to bytes."""
+    import cv2
+
+    # pylint: disable=no-member
+    encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 90]
+    # pylint: disable=no-member
+    success, data = cv2.imencode('.jpg', cv_image, encode_param)
+
+    if success:
+        return data.tobytes()
+
+    return None
+
+
+def cv_image_from_bytes(image):
+    """Convert image bytes to OpenCV image."""
+    import cv2
+    import numpy
+
+    # pylint: disable=no-member
+    return cv2.imdecode(numpy.asarray(bytearray(image)), cv2.IMREAD_UNCHANGED)
+
+
+def process_image(image, classifier_group, is_camera):
+    """Process the image given a classifier group."""
+    import cv2
+    import numpy
+
+    # pylint: disable=no-member
+    cv_image = cv2.imdecode(numpy.asarray(bytearray(image)),
+                            cv2.IMREAD_UNCHANGED)
+    group_matches = {}
+    for classifier_config in classifier_group:
+        classifier_path = classifier_config[CONF_FILE_PATH]
+        classifier_name = classifier_config[CONF_NAME]
+        color = classifier_config[CONF_COLOR]
+        scale = classifier_config[CONF_SCALE]
+        neighbors = classifier_config[CONF_NEIGHBORS]
+        min_size = classifier_config[CONF_MIN_SIZE]
+
+        # pylint: disable=no-member
+        classifier = cv2.CascadeClassifier(classifier_path)
+
+        detections = classifier.detectMultiScale(cv_image,
+                                                 scaleFactor=scale,
+                                                 minNeighbors=neighbors,
+                                                 minSize=min_size)
+        regions = []
+        # pylint: disable=invalid-name
+        for (x, y, w, h) in detections:
+            if is_camera:
+                # pylint: disable=no-member
+                cv2.rectangle(cv_image,
+                              (x, y),
+                              (x + w, y + h),
+                              color,
+                              2)
+            else:
+                regions.append((int(x), int(y), int(w), int(h)))
+        group_matches[classifier_name] = regions
+
+    if is_camera:
+        return cv_image_to_bytes(cv_image)
+    else:
+        return group_matches
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up the OpenCV platform entities."""
+    _LOGGER.info('Async setup for opencv')
+    if not os.path.isfile(DEFAULT_CLASSIFIER_PATH):
+        _LOGGER.info('Downloading default classifier')
+        import urllib3
+
+        http = urllib3.PoolManager()
+        request = http.request('GET', CASCADE_URL, preload_content=False)
+
+        with open(DEFAULT_CLASSIFIER_PATH, 'wb') as out:
```

Comment:
Oh you're right, I missed that.

@Teagan42 we should not do any I/O inside an asynchronous context without yielding. See [Async 101](https://home-assistant.io/developers/asyncio_101/)